add-controllers

pankovrec opened the pull request:
1st commit

pankovrec committed:
add-controllers
1st commit

utausheva commented on src/main/java/ru/practicum/shareit/exceptions/ErrorHandler.java:
```diff
@@ -0,0 +1,25 @@
+    @ExceptionHandler
+    public ResponseEntity<Error> notFoundItemException(NotFoundItemException e) {
+        return new ResponseEntity<>(new Error(HttpStatus.NOT_FOUND.value(), e.getMessage()), HttpStatus.NOT_FOUND);
+    }
```
1) добавь еще метод для остальных непредвиденных исключений - для Throwable)
2) на этапе обучения было бы круто писать лог с текстом ошибки, чтобы научиться разбираться в том, что подсказывает тебе java)

pankovrec replied:
Привет. Поправил здесь.

utausheva commented on src/main/java/ru/practicum/shareit/user/model/User.java:
```diff
@@ -0,0 +1,23 @@
+    @Email
+    private String email;
+
+    public User() {
```
можно заменить на @NoArgsConstructor

pankovrec replied:
здесь тоже заменил

utausheva commented on src/main/java/ru/practicum/shareit/user/UserStorageImpl.java:
```diff
@@ -0,0 +1,82 @@
+
+    @Override
+    public User addUser(User user) {
+        if (validateEmail(user)) {
```
по смыслу названия метода получается, что если валидация ок, то кидается исключение :D это не страшно, на следующем спринте такого метода уже не будет)

pankovrec replied:
не стал ничего с этим делать.

utausheva commented on src/main/java/ru/practicum/shareit/user/UserStorageImpl.java:
```diff
@@ -0,0 +1,82 @@
+        }
+        if (user.getEmail() != null) {
+            if (validateEmail(user)) {
+                throw new DuplicateEmailException(
```
не очень хорошо выглядит такая запись исключения, оно ведь вполне помещается в одну строку, максимум в 2)

pankovrec replied:
исправил

utausheva commented on src/main/java/ru/practicum/shareit/user/UserStorageImpl.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    private Long makeId() {
+        return (id == 0L) ? id = 1L : ++id;
```
а нужна ли здесь такая проверка? 0L будет только на старте, а ты заставляешь java каждый раз проверять

pankovrec replied:
убрал

utausheva commented on src/main/java/ru/practicum/shareit/user/UserService.java:
```diff
@@ -0,0 +1,18 @@
+import java.util.Collection;
+
+public interface UserService {
+    User createUser(User user);
```
давай в сервис тоже добавим описания, как в хранилище? Выглядит такой код очень выигрышно)

pankovrec replied:
сделал

utausheva commented on src/main/java/ru/practicum/shareit/user/UserService.java:
```diff
@@ -0,0 +1,18 @@
+public interface UserService {
+    User createUser(User user);
+
+    User updateUser(long userId, User user);
```
кстати, возвращать сервис должен дто, а не модель, во всех методах

pankovrec replied:
готово!

utausheva commented on src/main/java/ru/practicum/shareit/user/UserServiceImpl.java:
```diff
@@ -0,0 +1,43 @@
+
+    @Override
+    public User updateUser(long userId, User user) {
+        return userStorage.updateUser(userId, user);
```
не хватает проверки на существование юзера по ид в хранилище

pankovrec replied:
сделал

utausheva commented on src/main/java/ru/practicum/shareit/user/UserServiceImpl.java:
```diff
@@ -0,0 +1,43 @@
+
+    @Override
+    public UserDto getUser(long userId) {
+        return userStorage.getUser(userId);
```
тоже надо сначала проверить, есть ли пользователь в хранилище, будет полезно для следующего спринта

pankovrec replied:
сделал

utausheva commented on src/main/java/ru/practicum/shareit/user/UserController.java:
```diff
@@ -1,12 +1,47 @@
-}
+    private final UserService userService;
+
+    public UserController(UserService userService) {
```
а в сервисе написал Autowired, а здесь нет) Забыл или не знал?)

pankovrec replied:
забыл))

utausheva commented on src/main/java/ru/practicum/shareit/item/ItemStorageImpl.java:
```diff
@@ -0,0 +1,101 @@
+                items.get(userId).add(newItem);
+                return ItemMapper.toItemDto(newItem);
+            } else {
+                throw new NotFoundItemException(String.format("У пользователя %s нет вещи с id = %s",
```
давай чтобы избавиться от такой дикой вложенности сначала проверим items.get(userId) на null и бросим исключение, а потом уже вне ифов делать остальные проверки)

pankovrec replied:
избавился.

utausheva reviewed (commented):
Привет!
Работа хорошая, никаких критических замечаний нет, только рекомендации для улучшения качества кода))
Думаю, справишься быстро) Если будут вопросы - пиши в слак или в пачку ;)

pankovrec committed:
add-controllers
1st commit

pankovrec committed:
add-controllers
refactor service возвращает dto
style внесены поправки согласно ревью
feat проработаны прежние недоработки по исключениям.

pankovrec committed:
add-controllers
refactor service возвращает dto
style внесены поправки согласно ревью
feat проработаны прежние недоработки по исключениям.

utausheva commented on src/main/java/ru/practicum/shareit/user/UserStorage.java:
```diff
@@ -0,0 +1,34 @@
+    /**
+     * добавление пользователя
+     */
+    UserDto addUser(User user);
```
возможно, упустила на прошлой итерации, но хранилища должны отдавать модель, не дто. Преобразование модели в дто происходит на уровне сервиса.

utausheva reviewed (approved):
Супер!
Только подправь, пожалуйста, хранилища, они должны возвращать сущности модели, а не дто.. Думаю, это не составит труда и ты поправишь самостоятельно)
После правок можешь мерджить))
Если будут вопросы - пиши в пачку или здесь, я увижу))
